ryanoasis/nerd-fonts
# Improve CI process for releases

While preparing the `2.2.0` release I found several shortcomings in the current process.
* #881

This is a loose collection of all the changes that resulted in the process.
There are a lot changes, but I tried to break them down into digestible chunks / commits.

#### Requirements / Checklist


#### What does this Pull Request (PR) do?

* We can have releases and release candidates
* Releases are immutable. Release candidates will be updated on any push to `master`
* Both types are initiated by bumping the version in `package.json` and pushing it to `master`
* Release tags start with `v` (e.g. `v2.3.0-RC` or `v3.0.0`) (but not the number in `package.json`)

The release process will:
1. Update the versions in all scripts
1. Patch all fonts (that are mentioned in `bin/scripts/lib/fonts.json`)
1. Create single font archives on the release page (e.g. `Agave.zip`) **[\*]**
2. Create `cask`s for the archives (as artifact, not further processed)
3. Create the `font-patcher.zip` archive on the release page **[\*]**
4. Update the fontconfig `10-nerd-font-symbols.conf`
5. Commit the patched fonts back to `patched-fonts/`
6. Commit fontconfig back to repo (if changed)
7. Commit version-changed scripts back to repo (if changed)
8. Move release git-tag to include 7, 8, and 9 **[\*]**

[*] Note that steps 3, 5, and 10 are **ommitted** when there is no release candidate (no running release) and we already have a (immutable) release. Then just the files in the repo are updated, but nothing can be seen on the release page.

#### How should this be manually tested?

Apart from the checks that the release files contain useful data the logic behind the releases has to be tested:

1. Create a new (real) release
2. Check the release files
3. Push something to master, repo is updated, release itself is not changed
4. Create a new release candidate
2. Check the release files
6. Push something to master, repo and release is updated, release tag shifted

Note that the tests need to be done on the `master` branch to check the automatic workflow triggers.
Probably a fork is a good solution.

#### Any background context you can provide?
* #881 

#### What are the relevant tickets (if any)?

#### Screenshots (if appropriate or helpful)

Test as described above (with a small font subset).

**Step 1**

Create a new release by:
```diff
commit fc925f59849b4c6de002ba743ffe6fc583bfec22
Refs: 2.2.0-RC_old-229-gfc925f598
Author:     Fini Jastrow <ulf.fini.jastrow@desy.de>
AuthorDate: Fri Aug 26 09:59:28 2022 +0200
Commit:     Fini Jastrow <ulf.fini.jastrow@desy.de>
CommitDate: Fri Aug 26 09:59:28 2022 +0200

    simulate release 723
---
 package.json | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/package.json b/package.json
index 3daf22a98..3be7f9d9f 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "nerd-fonts",
-  "version": "2.2.0-RC",
+  "version": "7.3.2",
   "description": "Iconic font aggregator, collection, & patcher. 3,600+ icons, 50+ patched fonts: Hack, Source Code Pro, more. Glyph collections: Font Awesome, Material Design Icons, Octicons, & more",
   "main": "index.js",
   "dependencies": {},
```

https://github.com/Finii/nerd-fonts/actions/runs/2932178246

![image](https://user-images.githubusercontent.com/16012374/186881789-26f84fc0-b8e2-47d2-814b-75fc749ea676.png)

Note the outputs that show exactly what kind of release has been detected and if it is immutable or not:

![image](https://user-images.githubusercontent.com/16012374/186881980-3a117acd-ebfe-4c08-99f7-3b00affe0180.png)

**Step 2**

![image](https://user-images.githubusercontent.com/16012374/186882205-b377a196-3b8f-4714-a88c-7173df9f2884.png)

**Step 3**

```diff
commit c9d4965465ffe1bfe185d53da7a169d0f9ceec4d
Refs: 2.2.0-RC_old-233-gc9d496546
Author:     Fini Jastrow <ulf.fini.jastrow@desy.de>
AuthorDate: Fri Aug 26 10:18:24 2022 +0200
Commit:     Fini Jastrow <ulf.fini.jastrow@desy.de>
CommitDate: Fri Aug 26 10:19:22 2022 +0200

    push to master, refresh repo
---
 font-patcher | 1 +
 1 file changed, 1 insertion(+)

diff --git a/font-patcher b/font-patcher
index ab0ee5180..7cc6e8ee3 100755
--- a/font-patcher
+++ b/font-patcher
@@ -3,6 +3,7 @@
 # Nerd Fonts Version: 7.3.2
 # Script version is further down

+
 from __future__ import absolute_import, print_function, unicode_literals

 # Change the script version when you edit this script:
```

https://github.com/Finii/nerd-fonts/actions/runs/2932292079

![image](https://user-images.githubusercontent.com/16012374/186882944-d9073c08-1f3a-4ed0-8cc5-4a80d7fcc1b7.png)

Some steps are disabled now, for example:

![image](https://user-images.githubusercontent.com/16012374/186883225-060b2e35-2e68-4fdb-898d-4690394d021e.png)

Checking the release page, it is unchanged. Well the files are not updated. But the repo is:

![image](https://user-images.githubusercontent.com/16012374/186883940-171b820e-e43b-4f21-a3ee-ed1a9fcf7691.png)

**Step 4**

```diff
commit 3b4dd7a9f5566a4e8973684093f223f624940f5e
Refs: 2.2.0-RC_old-235-g3b4dd7a9f
Author:     Fini Jastrow <ulf.fini.jastrow@desy.de>
AuthorDate: Fri Aug 26 10:48:24 2022 +0200
Commit:     Fini Jastrow <ulf.fini.jastrow@desy.de>
CommitDate: Fri Aug 26 10:48:24 2022 +0200

    simulate release 859-RC

    Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>
---
 package.json | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/package.json b/package.json
index 3be7f9d9f..6f05b6a91 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "nerd-fonts",
-  "version": "7.3.2",
+  "version": "8.5.9-RC",
   "description": "Iconic font aggregator, collection, & patcher. 3,600+ icons, 50+ patched fonts: Hack, Source Code Pro, more. Glyph collections: Font Awesome, Material Design Icons, Octicons, & more",
   "main": "index.js",
   "dependencies": {},
```

https://github.com/Finii/nerd-fonts/actions/runs/2932455976

![image](https://user-images.githubusercontent.com/16012374/186884485-05c85949-e163-4f4c-a226-7b6c0d0dfd49.png)

**Step 5**

![image](https://user-images.githubusercontent.com/16012374/186884814-cee103bd-f883-4d8e-bff3-5ed9e54fb1d9.png)
_Note: This is not the correct image, it is in fact taken after Step 6_

**Step 6**

```diff
commit ce09410aa8b5d6f58a397b7ad7f2fe77c3a28e0d
Refs: [master], 2.2.0-RC_old-238-gce09410aa
Author:     Fini Jastrow <ulf.fini.jastrow@desy.de>
AuthorDate: Fri Aug 26 11:14:38 2022 +0200
Commit:     Fini Jastrow <ulf.fini.jastrow@desy.de>
CommitDate: Fri Aug 26 11:14:38 2022 +0200

    push to master, refresh repo
---
 font-patcher | 1 -
 1 file changed, 1 deletion(-)

diff --git a/font-patcher b/font-patcher
index 38ef3cdce..1b8bf994c 100755
--- a/font-patcher
+++ b/font-patcher
@@ -3,7 +3,6 @@
 # Nerd Fonts Version: 8.5.9-RC
 # Script version is further down

-
 from __future__ import absolute_import, print_function, unicode_literals

 # Change the script version when you edit this script:
```

https://github.com/Finii/nerd-fonts/actions/runs/2932617566

![image](https://user-images.githubusercontent.com/16012374/186885206-9e761a40-973a-4126-8533-99cecfdd96cc.png)

The rolling release / release candidate is shifted,. The circled items will show different values than before:

![image](https://user-images.githubusercontent.com/16012374/186885560-23585d9a-7084-4a6c-b546-abd67b8f207b.png)

Example git log (not complete)
![image](https://user-images.githubusercontent.com/16012374/186885742-e749efd1-6597-4e29-8549-9eae8a22be43.png)
